version.md

```diff
@@ -0,0 +1,78 @@
+# Versioning endpoints
+
+[Back to the list of all defined endpoints](endpoints.md)
+
+This endpoint represents a single version, matching a single item.
+
+Whether a version is accessible depends on `versioning.item.history.view.admin` configuration.  
+If this is set to true, the version can only be retrieved if the user is an admin of the related item
+
+## Create version
+
+TODO
+
+## Get single version
+
+**GET /api/versioning/version/<:versionId>**
+
+Provide version information for the version id.
+
+```json
+{
+  "id": "102",
+  "version": "2",
+  "type": "version",
+  "created": "2015-11-03T09:44:46.617",
+  "summary": "Fixing some typos in the abstract",
+  "_links": {
+    "versionhistory": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/versioning/versionhistory/10"
+    },
+    "self": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/versioning/version/101"
+    },
+    "eperson": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/versioning/version/101/eperson"
+    },
+    "item": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/versioning/version/101/item"
+    }
+  }
+}
+```
+
+Status codes:
+* 200 OK - if the version exists and is accessible by the current user
+* 401 Unauthorized - if you are not authenticated and versioning is not public
+* 403 Forbidden - if you are not logged in with sufficient permissions and versioning is not public
+* 404 Not found - if the version doesn't exist
+
+## Get single version for item
+
+**GET /api/versioning/item/<:item-uuid>**
```

Comment:
ah ok, I started providing feedback https://github.com/DSpace/Rest7Contract/pull/97#discussion_r376516883 in the conversation tab and I haven't understand the goal.

This is a search method for the versions endpoint so it should be
`/api/versioning/versions/search/item?uuid=<:uuid>` or `/api/versioning/itemversions/search/item?uuid=<:uuid>`

Reply:
I talked briefly to @abollini about this via Slack and pointed out that this search would always return a unique result (as each version is always represented by a unique Item). 

He pointed out that another option here would be to treat this endpoint as a *redirect*, instead of a search.   So, we could keep a "lookup" endpoint of : `/api/versioning/item/<:item-uuid>` but note that it always redirects to `/api/versioning/versions/{:id}` -- as it's simply a different path to looking up a unique version.  This is in more line with how `/api/pid/find` endpoint works...it redirects you to the endpoint where the resource matching that unique ID can be found.

Reply:
This syntax is actually based on https://stackoverflow.com/questions/20381976/rest-api-design-getting-a-resource-through-rest-with-different-parameters-but/20386425#20386425, but there are no official guidelines on how to approach this.

Using a redirect is also possible if it's not used frequently or if it's very lightweight because it implies:
* REST will need to load all information to return the redirect
* Angular will need to request the actual details based on the redirect
* REST will need to load all information again to return the actual version info

For performance reasons, I do prefer the `GET /api/versioning/item/<:item-uuid>` and it's not uncommon practice

Reply:
@benbosman : I had talked with @abollini about this a bit on Slack on Friday.  I think there's a small difference of opinion here with the approach laid out in that StackOverflow article you linked to.

Andrea noted his belief that each "resource" should have one primary URI. For our purposes, that primary URI is via the UUID (e.g. `/api/versioning/versions/{:id}` is the primary URI for a single version).  If we were to also have `/api/versioning/items/{:item-uuid}` return the **same resource** (i.e. same JSON response), then this would mean these version resources would now have two primary URIs. 

So, Andrea was suggesting that either this URL could redirect to the primary URI (which, as you pointed out may have some performance issues), or it could be achieved via a search, where the result set (of one version) points back at the primary URI.  In other words, Andrea seems to agree more with the *second answer* in that same StackOverflow question: https://stackoverflow.com/a/20388766

I'm also now realizing that this URI is slightly confusing as-is.  The main resource here is implied to be an `item` (as our URIs have the structure `/api/{category}/{resource}`....so this URL claims to return an `item` resource, and not a `version` or `itemversion` resource.)  So, to correctly structure this URI would be to make it something like `/api/versioning/versions/item/<:item-uuid>`, which looks a bit odd to me.

I'm personally wondering if we should just start using *searches* for anything that is NOT a UUID (or unique ID of another form).  The code underneath the search URI can still be a simple lookup (rather than a Discovery search), if we know that field is always unique.  But, at least it makes it clearer that UUIDs are the identifier we prefer clients to use...rather than other unique fields.  Thoughts here?

(Sidenote: @benbosman : I think this Contract is honestly mostly OK as-is, so implementation work need not be held up.  We all also agree this sort of "lookup" by Item UUID is needed...the only area of disagreement is around the URI path to use.)

Reply:
Perhaps the confusion is that this is not a primary URI.
Would it be better to use `/api/core/items/{:item-uuid}/version` similar to `/api/core/items/{:item-uuid}/owningCollection`. It's just an alternative URL to get to the version based on the item UUID

Changing the path can indeed still happen during the implementation

Reply:
@benbosman : I'm OK with `/api/core/items/{:item-uuid}/version` as it makes it much clearer that this is not a primary URI for the Version resource.  To me, that clarifies that we are obtaining a Version via its relationship to a known Item.  Good suggestion!

Reply:
it is fine to me too to add `/api/core/items/{:item-uuid}/version`, Nevertheless, this has a slightly different meaning than a search endpoint or a redirect.
If we have this subpath in the item endpoint to be adherent to the HATEOAS principles we should use it only after having retrieved an actual item, i.e. we are looking to an item with such link we know how to move to the version. If we want to implement a search form where you can find the version by item uuid it should be a search method, if we plan to have a lookup form where we can put an alternative identifier for the version item (the item uuid) we should use the redirection.
BTW, we have just discovered that the search methods already support returning a single value and it is already used in the existing implementation of the EPerson endpoint but unfortunately this was not documented in the rest contract, see
https://github.com/DSpace/DSpace/blob/a0ca6e2fadb60e718d60685ae4f9a7ed1efb2101/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java#L156

(a PR to document this endpoint and fix the javadoc is coming as part of the work on DS-4428 indeed we discover this endpoint as it was failing the test after the fix)

Reply:
@abollini and @benbosman : To clarify my point here, my assumption was also that the endpoint `/api/core/items/{:item-uuid}/version` would essentially be a "link" from an Item to a Version (just like `/owningCollection` is a "link" from the Item to its Collection and *not* a way to search all Collections via an Item ID).

So, `/api/core/items/{:item-uuid}/version` would be like the `Item` or `ItemService` class having a "getVersion()` method on it.  But, if we are really looking to model a *query* across all Versions based on an Item UUID, then I agree with @abollini that it likely is better to just do this as a one-result search.

I'm honestly fine with either (or both) of these endpoints existing, but I don't know what the initial use case was that @benbosman was looking to address.  Is this a "search endpoint" (essentially a `SELECT * from versionitem where item_id={uuid}`)?  Or is this a "link" between an already loaded Item object and the Version it represents?  The former should be more like `/versions/search/item?uuid={:uuid}` while the latter could be `/item/{:uuid}/version` as proposed by Ben.

Reply:
The endpoint would use [VersioningService.getVersion()](https://github.com/DSpace/DSpace/blob/a0ca6e2fadb60e718d60685ae4f9a7ed1efb2101/dspace-api/src/main/java/org/dspace/versioning/service/VersioningService.java#L55)

I can see https://github.com/DSpace/DSpace/blob/a0ca6e2fadb60e718d60685ae4f9a7ed1efb2101/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java#L156 is indeed also returning a single EPerson instead of a list containing only one value (I wasn't aware this existed).

For this particular use-case, it would be used to display the current item's version, so it's not really intended to be a search method. The DSpace 6 UI is also based on an item page:
<img width="874" alt="Schermafdruk 2020-02-12 09 25 25" src="https://user-images.githubusercontent.com/1528033/74316397-c0d83d80-4d79-11ea-8f4e-85acf99758d5.png">

Based on this, I will modify the endpoint to /api/core/items/{:item-uuid}/version

Reply:
yes, according to this use case I'm fine with the subpath /version